### DISCUSSION TITLE
Is there defined ordering in multiple declarations of partial record?

### DISCUSSION THREAD

**Discussion by ufcpp** [dotnet/csharplang#4957] [General] [answered]
As a comparison, having fields in multiple declarations of struct results in a warning CS0282.

```cs
partial struct Struct // Warning: CS0282, There is no defined ordering between fields in multiple declarations of partial struct
{
    public int X;
}

partial struct Struct
{
    public int Y;
}
```

Don't records need the same type of warning for their properties? Ordering of properties in records has a visible effect on the synthesized methods such as `ToString`.

```cs
// in the same file
partial record class Record
{
    public int X { get; init; }
}

partial record class Record
{
    public int Y => X * X;
}
// WriteLine(new Record { X = 2 });
// Record { X = 2, Y = 4 }
```

```cs
// in the same file
partial record class Record
{
    public int Y => X * X;
}

partial record class Record
{
    public int X { get; init; }
}
// WriteLine(new Record { X = 2 });
// Record { Y = 4, X = 2 }
```

```cs
// A.cs
partial record class Record
{
    public int X { get; init; }
}

// B.cs
partial record class Record
{
    public int Y => X * X;
}
// WriteLine(new Record { X = 2 });
// Record { X = 2, Y = 4 }
```

```cs
// A.cs
partial record class Record
{
    public int Y => X * X;
}

// B.cs
partial record class Record
{
    public int X { get; init; }
}
// WriteLine(new Record { X = 2 });
// Record { Y = 4, X = 2 }
```

Should this behavior be defined in the spec, especially for cross-file partial declarations?

**Comment by RikkiGibson** [comment, marked as answer]
It seems like this warning is specific to structs since it more frequently interacts with issues of field layout in memory. We do give this warning with record structs. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmOATEBqAPgAQCYBGAWACgKAHCAJwSggBsACWgUwGMB7WjFlAloBXTghYAlLhQDeFFgpb4AzC1jioLGSwDm7BAG4B+owF8K5yuRr1GrDjz4Cho8VM6z5ilWpjiAVlq6JsaGLJamQA=).

I think it's probably fine to explicitly say in the spec that an implementation-defined ordering is used when determining the order in which to call printable members in ToString for record classes and record structs, for example. But it doesn't seem worthy of a warning to me.

cc @jcouv for visibility.

**Reply by jcouv** [reply]
It's not a new problem or specific to records, but it seems surprising that the [spec](https://github.com/dotnet/csharplang/blob/main/spec/classes.md#partial-types) doesn't call out the order in which parts of a partial type are stitched together. 
The compiler certainly has an order. If I recall correctly, it's based on the syntactic order and order of inputs given to the compilation.

@RikkiGibson Do you want to start a thread with LDM?
